File: web/src/lib/api/cyberchef.ts
/**
 * Client-side API helpers for the CyberChef integration endpoints.
 * These functions call the SvelteKit server proxy routes which in turn
 * forward requests to the nettap-storage-daemon.
 */

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface CyberChefStatus {
	available: boolean;
	version: string;
	container_running: boolean;
	container_name: string;
}

export interface CyberChefRecipe {
	name: string;
	description: string;
	category: string;
	recipe_fragment: string;
}

// ---------------------------------------------------------------------------
// Fetch helpers
// ---------------------------------------------------------------------------

/**
 * Get CyberChef container availability and version information.
 */
export async function getCyberChefStatus(): Promise<CyberChefStatus> {
	const res = await fetch('/api/cyberchef/status');

	if (!res.ok) {
		return {
			available: false,
			version: '',
			container_running: false,
			container_name: '',
		};
	}

	return res.json();
}

/**
 * Retrieve the list of pre-built CyberChef recipes.
 * Optionally filter by category (e.g. "decode", "encode", "network").
 */
export async function getRecipes(
	category?: string
): Promise<{ recipes: CyberChefRecipe[]; count: number }> {
	const params = category ? `?category=${encodeURIComponent(category)}` : '';
	const res = await fetch(`/api/cyberchef/recipes${params}`);

	if (!res.ok) {
		return { recipes: [], count: 0 };
	}

	return res.json();
}

/**
 * Build a CyberChef URL with the given recipe fragment and input data.
 * Returns a URL that can be opened in an iframe or new tab.
 */
export async function buildRecipeUrl(
	recipe_fragment: string,
	input_data: string
): Promise<{ url: string }> {
	const res = await fetch('/api/cyberchef/url', {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify({ recipe_fragment, input_data }),
	});

	if (!res.ok) {
		const body = await res.json().catch(() => ({}));
		throw new Error(body.error || `Failed to build recipe URL (HTTP ${res.status})`);
	}

	return res.json();
}

File: daemon/api/cyberchef.py
"""
NetTap CyberChef API Routes

Registers CyberChef-related endpoints with the aiohttp application.
"""

import logging
from aiohttp import web

from services.cyberchef_service import CyberChefService

logger = logging.getLogger("nettap.api.cyberchef")


async def handle_cyberchef_status(request: web.Request) -> web.Response:
    """GET /api/cyberchef/status — CyberChef container status."""
    cyberchef: CyberChefService = request.app["cyberchef"]
    try:
        status = await cyberchef.is_available()
        return web.json_response(status)
    except Exception as exc:
        logger.exception("Failed to check CyberChef status")
        return web.json_response({"error": str(exc)}, status=500)


async def handle_cyberchef_recipes(request: web.Request) -> web.Response:
    """GET /api/cyberchef/recipes?category=decode — List pre-built recipes."""
    category = request.query.get("category", "")
    cyberchef: CyberChefService = request.app["cyberchef"]
    try:
        recipes = cyberchef.get_recipes(category)
        return web.json_response({"recipes": recipes, "count": len(recipes)})
    except Exception as exc:
        logger.exception("Failed to get recipes")
        return web.json_response({"error": str(exc)}, status=500)


async def handle_cyberchef_url(request: web.Request) -> web.Response:
    """POST /api/cyberchef/url — Build a CyberChef URL with recipe and input data."""
    try:
        body = await request.json()
    except Exception:
        return web.json_response({"error": "Invalid JSON body"}, status=400)

    cyberchef: CyberChefService = request.app["cyberchef"]
    recipe_fragment = body.get("recipe_fragment", "")
    input_data = body.get("input_data", "")

    if not recipe_fragment:
        return web.json_response({"error": "recipe_fragment is required"}, status=400)

    try:
        url = cyberchef.build_recipe_url(recipe_fragment, input_data)
        return web.json_response({"url": url})
    except Exception as exc:
        logger.exception("Failed to build CyberChef URL")
        return web.json_response({"error": str(exc)}, status=500)


def register_cyberchef_routes(
    app: web.Application, cyberchef: CyberChefService
) -> None:
    """Register CyberChef API routes on the aiohttp application."""
    app["cyberchef"] = cyberchef
    app.router.add_get("/api/cyberchef/status", handle_cyberchef_status)
    app.router.add_get("/api/cyberchef/recipes", handle_cyberchef_recipes)
    app.router.add_post("/api/cyberchef/url", handle_cyberchef_url)
    logger.info("CyberChef API routes registered (3 endpoints)")
